File: benchmarks/semantic_search/fixtures/python-workflow-discovery-v1/discovery/catalog.py
from __future__ import annotations

from dataclasses import dataclass, field


@dataclass
class Workflow:
    workflow_id: str
    parameters: dict[str, str] = field(default_factory=dict)
    labels: dict[str, str] = field(default_factory=dict)


class WorkflowCatalog:
    def __init__(self, workflows: list[Workflow]) -> None:
        self.workflows = {workflow.workflow_id: workflow for workflow in workflows}

    def list_workflows(self, filters: dict[str, str] | None = None) -> list[Workflow]:
        workflows = list(self.workflows.values())
        if not filters:
            return workflows
        return [workflow for workflow in workflows if all(workflow.labels.get(k) == v for k, v in filters.items())]

    def list_actions(self, filters: dict[str, str]) -> list[Workflow]:
        return self.list_workflows(filters)

    def get_workflow(self, workflow_id: str) -> Workflow | None:
        return self.workflows.get(workflow_id)

File: benchmarks/semantic_search/fixtures/go-workflow-discovery-v1/internal/discovery/catalog.go
package discovery

import "context"

type Workflow struct {
	ID         string
	Parameters map[string]string
	Labels     map[string]string
}

type WorkflowCatalog struct {
	workflows map[string]Workflow
}

func (c *WorkflowCatalog) ListWorkflows(ctx context.Context, filters DiscoveryFilters) []Workflow {
	_ = ctx
	_ = filters
	result := make([]Workflow, 0, len(c.workflows))
	for _, workflow := range c.workflows {
		result = append(result, workflow)
	}
	return result
}

// ListActions applies signal-derived filters before exposing available actions.
func (c *WorkflowCatalog) ListActions(ctx context.Context, filters DiscoveryFilters) []Workflow {
	return ApplyLabelFilters(ctx, c.ListWorkflows(ctx, filters), filters)
}

func (c *WorkflowCatalog) GetWorkflow(ctx context.Context, workflowID string) (Workflow, bool) {
	_ = ctx
	workflow, ok := c.workflows[workflowID]
	return workflow, ok
}
